core/src/com/google/inject/internal/aop/AbstractGlueGenerator.java

```diff
@@ -0,0 +1,214 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.inject.internal.aop;
+
+import static java.lang.reflect.Modifier.PUBLIC;
+import static java.lang.reflect.Modifier.STATIC;
+import static org.objectweb.asm.Opcodes.ACONST_NULL;
+import static org.objectweb.asm.Opcodes.ARETURN;
+import static org.objectweb.asm.Opcodes.F_SAME;
+import static org.objectweb.asm.Opcodes.ILOAD;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Executable;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.SortedMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.ToIntFunction;
+import org.objectweb.asm.ClassWriter;
+import org.objectweb.asm.Label;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Type;
+
+/**
+ * Support code for generating enhancer/fast-class glue.
+ *
+ * <p>Each glue class has a trampoline that accepts an index, context object, and argument array:
+ *
+ * <pre>
+ * public static Object GUICE$TRAMPOLINE(int index, Object context, Object[] args) {
+ *   switch (index) {
+ *     case 0: {
+ *       return ...;
+ *     }
+ *     case 1: {
+ *       return ...;
+ *     }
+ *   }
+ *   return null;
+ * }
+ * </pre>
+ *
+ * Each indexed statement in the trampoline invokes a constructor or method, returning the result.
+ * The expected context object depends on the statement; it could be the invocation target, some
+ * additional constructor context, or it may be unused. Arguments are unpacked from the array onto
+ * the call stack, unboxing or casting them as necessary. Primitive results are autoboxed before
+ * being returned.
+ *
+ * <p>Where possible the trampoline is converted into a lookup {@link Function} mapping an integer
+ * to an invoker function, each invoker represented as a {@link BiFunction} that accepts a context
+ * object plus argument array and returns the result. These functional interfaces are used to avoid
+ * introducing a dependency from the glue class to Guice specific types. This means the glue class
+ * can be loaded anywhere that can see the host class, it doesn't need access to Guice's own {@link
+ * ClassLoader}. (In other words it removes any need for bridge {@link ClassLoader}s.)
+ *
+ * @author mcculls@gmail.com (Stuart McCulloch)
+ */
+abstract class AbstractGlueGenerator {
+
+  protected static final String GENERATED_SOURCE = "<generated>";
+
+  protected static final String TRAMPOLINE_NAME = "GUICE$TRAMPOLINE";
+
+  /**
+   * The trampoline method takes an index, along with a context object and an array of argument
```

Comment:
the type descriptor does not include this index parameter?

Reply:
it does:
```
"(ILjava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;";
  ^
```